src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Microsoft.CodeAnalysis.MSBuild.Logging;
+using Roslyn.Utilities;
+using MSB = Microsoft.Build;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Build
+{
+    internal class ProjectBuildManager
+    {
+        private static readonly XmlReaderSettings s_xmlReaderSettings = new XmlReaderSettings()
+        {
+            DtdProcessing = DtdProcessing.Prohibit,
+            XmlResolver = null
+        };
+
+        private readonly MSB.Evaluation.ProjectCollection _projectCollection;
+        private readonly MSBuildDiagnosticLogger _logger;
+        private bool _started;
+
+        public ProjectBuildManager()
+        {
+            var properties = new Dictionary<string, string>()
+            {
+                // this will tell msbuild to not build the dependent projects
+                { PropertyNames.DesignTimeBuild, bool.TrueString },
+
+                // this will force CoreCompile task to execute even if all inputs and outputs are up to date
+                { PropertyNames.BuildingInsideVisualStudio, bool.TrueString },
+
+                { PropertyNames.BuildProjectReferences, bool.FalseString },
+                { PropertyNames.BuildingProject, bool.FalseString },
+
+                // retrieve the command-line arguments to the compiler
+                { PropertyNames.ProvideCommandLineArgs, bool.TrueString },
+
+                // don't actually run the compiler
+                { PropertyNames.SkipCompilerExecution, bool.TrueString },
+
+                { PropertyNames.ContinueOnError, PropertyValues.ErrorAndContinue }
+            };
+
+            _projectCollection = new MSB.Evaluation.ProjectCollection(properties);
+
+            _logger = new MSBuildDiagnosticLogger()
+            {
+                Verbosity = MSB.Framework.LoggerVerbosity.Normal
+            };
+        }
+
+        private MSB.Evaluation.Project FindProject(string path, IDictionary<string, string> globalProperties, CancellationToken cancellationToken)
+        {
+            var loadedProjects = _projectCollection.GetLoadedProjects(path);
```

Comment:
A HashSet? I guess you mean a map where the key is "project path" + "global properties the project was loaded with"? We already have an MSBuild `ProjectCollection` with the loaded projects contained within. So, (IMHO) we should just use it.

In addition, there are several subtleties of MSBuild global properties that we would likely have to replicate in order to do this correctly. Otherwise, subtle bugs might creep in:

* MSBuild's `GlobalProperties` is instantiated with a custom `MSBuildNameIgnoreCaseComparer`.
* Property values are escaped. (FWIW, I just realized that we weren't handling this fact, and will correct it now. :smile:)

I'm kind of loathe to try and manage that, so I've chosen to just use `ProjectCollection`. Does that work for you?